exec/java-exec/src/main/java/org/apache/drill/exec/planner/sql/handlers/CreateFunctionHandler.java

```diff
@@ -0,0 +1,195 @@
+import org.apache.drill.exec.store.sys.store.DataChangeVersion;
+import org.apache.drill.exec.util.JarUtil;
+import org.apache.drill.exec.work.foreman.ForemanSetupException;
+import org.apache.hadoop.fs.FileSystem;
+import org.apache.hadoop.fs.FileUtil;
+import org.apache.hadoop.fs.Path;
+
+import java.io.File;
+import java.io.IOException;
+import java.util.List;
+import java.util.UUID;
+
+public class CreateFunctionHandler extends DefaultSqlHandler {
+
+  private static org.slf4j.Logger logger = org.slf4j.LoggerFactory.getLogger(CreateFunctionHandler.class);
+
+  public CreateFunctionHandler(SqlHandlerConfig config) {
+    super(config);
+  }
+
+  /**
+   * Creates UDFs dynamically.
+   *
+   * @return - Single row indicating list of registered UDFs, raise exception otherwise
+   */
+  @Override
+  public PhysicalPlan getPlan(SqlNode sqlNode) throws ForemanSetupException, IOException {
+    if (context.getOption(ExecConstants.DYNAMIC_UDF_SUPPORT_ENABLED).bool_val) {
+      SqlCreateFunction node = unwrap(sqlNode, SqlCreateFunction.class);
+      String jarName = ((SqlCharStringLiteral) node.getJar()).toValue();
+      String sourceName = JarUtil.getSourceName(jarName);
+
+      RemoteFunctionRegistry remoteRegistry = context.getRemoteFunctionRegistry();
+      FileSystem fs = remoteRegistry.getFs();
+      Path tmpDir = new Path(remoteRegistry.getTmpArea(), UUID.randomUUID().toString());
+      File localTmpDir = Files.createTempDir();
+
+      boolean inProgress = false;
+      try {
+        final String action = remoteRegistry.addToJars(jarName, RemoteFunctionRegistry.Action.REGISTRATION);
+        if (!(inProgress = action == null)) {
+          return DirectPlan.createDirectPlan(context, false,
+              String.format("Jar with %s name is used. Action: %s", jarName, action));
+        }
+
+        // verify that binary and source exist
+        Path remoteBinary = new Path(remoteRegistry.getStagingArea(), jarName);
+        Path remoteSource = new Path(remoteRegistry.getStagingArea(), sourceName);
+        if (!fs.exists(remoteBinary) || !fs.exists(remoteSource)) {
+          return DirectPlan.createDirectPlan(context, false,
+              String.format("Binary [%s] or source [%s] is absent in udf staging area [%s].", jarName, sourceName, remoteRegistry.getStagingArea().toUri().getPath()));
+        }
+
+        // backup binary & source (copy to udf tmp directory)
+        fs.mkdirs(tmpDir);
+        Path tmpBinary = new Path(tmpDir, jarName);
+        Path tmpSource = new Path(tmpDir, sourceName);
+
+        FileUtil.copy(fs, remoteBinary, fs, tmpBinary, false, fs.getConf());
+        FileUtil.copy(fs, remoteSource, fs, tmpSource, false, fs.getConf());
+
+        // copy binary to local fs, we don't need source for validation
+        Path localBinary = new Path(new Path(localTmpDir.toURI()), jarName);
+        fs.copyToLocalFile(tmpBinary, localBinary);
+
+        // validate functions locally
```

Comment:
I don't immediately see how this flow avoids race conditions. There is no lock, so two threads in the same process can walk down these steps simultaneously.

Once a function is registered in ZK, how do we know that we are the client that did the registration (vs some other client at the same time.)

When we do registration, if we have a single node with all jars, how do we prevent race conditions?

ZK: a, b
Client X reads (a,b)
Client Y reads (a,b)
Client X writes (a,b,c)
Client Y writes (a,b,d)

In above, client X's c entry is overwritten. Or, do we have separate zNodes for each jar & function so we can atomically manage creates?

Where is the recovery code if jar j.jar has functions (f1, f2). We register f1 in ZK. But, f2 turns out to duplicate another fn just created by some other Drillbit. Where do we back out the f1 and j.jar registrations?

And so on.